src/utransport.rs

```diff
@@ -230,38 +240,47 @@ pub trait UTransport: Send + Sync {
     /// #     }
     /// # }
     /// #
-    /// # let my_transport = MyTransport::new();
-    /// # let my_uuri = UUri::default();
+    /// # let mut my_transport = MyTransport::new();
+    /// # let source_uuri = UUri::default();
     ///
     /// // hang onto this listener...
-    /// let my_listener: Arc<dyn UListener> = Arc::new(MyListener::new());
-    /// // ...send a clone through when registering...
-    /// let register_result = my_transport.register_listener(my_uuri.clone(), my_listener.clone());
+    /// let my_listener = Arc::new(MyListener::new());
+    /// // ...use a clone for registering...
+    /// my_transport.register_listener(
+    ///     Some(&source_uuri),
+    ///     None,
+    ///     my_listener.clone());
     /// // ...and use the original we hung onto when unregistering
-    /// let unregister_result = my_transport.unregister_listener(my_uuri, my_listener);
+    /// my_transport.unregister_listener(
+    ///     Some(&source_uuri),
+    ///     None,
+    ///     my_listener);
     /// ```
     async fn register_listener(
-        &self,
-        topic: UUri,
+        &mut self,
+        source_filter: Option<&UUri>,
+        sink_filter: Option<&UUri>,
```

Comment:
If I'm understanding correctly, the idea is rather than discriminate based on an explicit message type, to instead allow for implicit selection among the different types based on source and sink address.

I wanted to come back to something we discussed -- discriminating any Publish from any Notification.

I understand that the Notification will have a sink UUri.

Could you flesh out how we'd describe
1. "give me all Notifications (i.e. all messages whose sink is not empty)" vs 
2. "give me all Publishes (i.e. all messages whose sink is empty)"? 

Perhaps 1. we'd provide an empty UUri indicating any UUri filled will match? I'm not quite sure how we'd filter to get 2. as the semantics outlined is None => matches all UUris (including no UUri?).

Reply:
I think using an empty UUri to indicate any address would work. and then `None` truly means no UURI. I agree that there needs to be a distinction between `any` uuri and `no` uuri

Reply:
> I think using an empty UUri to indicate any address would work. and then None truly means no UURI. I agree that there needs to be a distinction between any uuri and no uuri

I second it. Perhaps we should also put the definition that empty UUri indicates any address into spec.

Reply:
@PLeVasseur I see your point.

This is very close to what I already specified in my UUri PR, i.e. using empty property values in a UUri to match any value.

@devkelley I believe this could work.

@evshary I will amend my UUri PR accordingly.

Reply:
> I am not sure if you really meant _Notification_ type messages in your first example or if you actually meant _all non-Publish messages_?

I meant all non-Publish, must have gotten mixed up.

Your examples makes sense. 👍